solver/app/cursor.go

```diff
@@ -0,0 +1,100 @@
+//nolint:unused // It will be used
+package app
+
+import (
+	"context"
+	"sync"
+
+	"github.com/omni-network/omni/lib/errors"
+	"github.com/omni-network/omni/lib/xchain"
+
+	ormv1alpha1 "cosmossdk.io/api/cosmos/orm/v1alpha1"
+	"cosmossdk.io/core/store"
+	"cosmossdk.io/orm/model/ormdb"
+	"cosmossdk.io/orm/types/ormerrors"
+	db "github.com/cosmos/cosmos-db"
+)
+
+// newSolverDB returns a new DB backend based on the given directory
+// or in-memory-based if the directory is empty.
+func newSolverDB(dbDir string) (db.DB, error) {
+	resp := db.DB(db.NewMemDB())
+	if dbDir != "" {
+		var err error
+		resp, err = db.NewGoLevelDB("solver", dbDir, nil)
+		if err != nil {
+			return nil, errors.Wrap(err, "new golevel db")
+		}
+	}
+
+	return resp, nil
+}
+
+func newCursors(db db.DB) (*cursors, error) {
+	schema := &ormv1alpha1.ModuleSchemaDescriptor{SchemaFile: []*ormv1alpha1.ModuleSchemaDescriptor_FileEntry{
+		{Id: 1, ProtoFileName: File_solver_app_solver_proto.Path()},
+	}}
+
+	storeSvc := dbStoreService{DB: db}
+
+	modDB, err := ormdb.NewModuleDB(schema, ormdb.ModuleDBOptions{KVStoreService: storeSvc})
+	if err != nil {
+		return nil, errors.Wrap(err, "create ormdb module db")
+	}
+
+	dbStore, err := NewSolverStore(modDB)
+	if err != nil {
+		return nil, errors.Wrap(err, "create store")
+	}
+
+	return &cursors{
+		table: dbStore.CursorTable(),
+	}, nil
+}
+
+// cursors provides a thread-safe persisted cursor store.
+type cursors struct {
+	mu    sync.Mutex
```

Comment:
Do we have to exclusively lock on reads?

Reply:
to be safe yes, the cosmosDB is a single threaded thing, I'm not sure concurrent reads are supported or not. We only read from the DB on startup in any case.